R/search.R

```diff
@@ -82,3 +102,165 @@ search <- function(query = NULL, source = NULL) {
 
   x[order(x$rank), ]
 }
+
+#' @title Creates a grid from grid parameters found in opendap.catalog::grids
+#' @description Creates a grid from grid parameters found in opendap.catalog::grids and then projects the grid to the CRS of input AOI. Internal helper function for the grid_subset function and is used in conjunction with the spatial_subset function.
+#' @param AOI SF object
+#' @param grid_data catalog data.frame from calling opendap.catalog::grids
+#' @details grid_data parameter is built to work on a single row of data from opendap.catalog::grids
+#' @return data.frame
+#' @export
+#' @importFrom magrittr `%>%`
+#' @importFrom sf st_bbox st_as_sfc st_crs st_transform st_as_sf
+
+fetch_grid <- function(AOI, grid_data) {
+
+
+  # X Coordinates
+  xcoords <- seq(
+    grid_data$X1,
+    grid_data$Xn,
+    length.out = grid_data$ncols
+  )
+
+  # # Check x coords
+  # if(any(xcoords > 180.001)) { xcoords = xcoords - 360}
+
+  if(grid_data$toptobottom == TRUE | is.na(grid_data$toptobottom)) {
+
+    # Grids going from "top to bottom"
+
+    # Y Coordinates
+    ycoords <- seq(
+      grid_data$Y1,
+      grid_data$Yn,
+      length.out = grid_data$nrows # by = grid_data$resY
+    )
+
+  } else if(grid_data$toptobottom == FALSE) {
+
+    # Grids going from "Bottom to Top"
+
+    # Y Coordinates
+    ycoords <- seq(
+      grid_data$Yn,
+      grid_data$Y1,
+      length.out = grid_data$nrows # by = grid_data$resY
+    )
+
+  }
+
+  # Turn off spherical geometry (s2)
+  sf::sf_use_s2(use_s2 = FALSE)
+
+  domain <-
+    sf::st_as_sfc(
+      sf::st_bbox(
+        c(
+          xmin = min(xcoords),
+          xmax = max(xcoords),
+          ymin = min(ycoords),
+          ymax = max(ycoords)
+        ),
+        crs =  sf::st_crs(grid_data$proj)
+      )
+    ) %>%
+    sf::st_transform(sf::st_crs(AOI)) %>%
+    sf::st_as_sf()
+
+  return(domain)
+
+}
+
+#' @title Check grids for intersection with AOI - Internal use
+#' @description Compares an input AOI and a grid created from the fetch_grid function and returns the grid ID number of any overlapping grids. Internal helper function for the grid_subset function and is used in conjunction with the fetch_grid function.
+#' @param AOI SF object
+#' @param grid_domain SF object representing spatial domain of grid, output of fetch_grid function
+#' @details Function is designed to take an AOI and compares it to the outputs from fetch_grid, returns the intersecting grid ID number as a character, and NULL if no intersection
+#' @return character
+#' @export
+#' @importFrom magrittr `%>%`
+#' @importFrom sf st_bbox st_as_sfc st_as_sf st_intersects
+
+spatial_subset <- function(AOI, grid_domain) {
+
+  aoi_bb   <-
```

Comment:
does this work better the getting the box from the AOI and filtering on the xmin, ... columns?